File: python/nexus_reasoner.py
"""
Darukaa.Earth - Python Multi-Metric Ecological Reasoning Engine
Evaluates >= 3 environmental variables simultaneously and computes 5-year restoration trajectories.
"""

from typing import Dict, Any, List
import math
from rag_pipeline import PythonEnvironmentalRAG

class PythonNexusReasoner:
    def __init__(self):
        self.rag = PythonEnvironmentalRAG()

    def evaluate(self, state: Dict[str, Any], query: str = "") -> Dict[str, Any]:
        soc = state.get("soil_organic_carbon")
        ph = state.get("soil_ph")
        rainfall_val = state.get("annual_rainfall", 350)
        crop = str(state.get("crop", "monoculture wheat")).lower()
        tillage = str(state.get("tillage", "conventional-inversion-tillage")).lower()

        nexus_insights = []
        bottlenecks = []
        interventions = []
        degradation_score = 30

        # Multi-variable coupling: SOC + Climate + Monoculture Crop
        if (soc is not None and soc < 1.0) or rainfall_val < 600 or "monoculture" in crop:
            if (soc is not None and soc <= 0.5) and rainfall_val < 500:
                nexus_insights.append({
                    "variables": [f"SOC ({soc}%)", f"Rainfall ({rainfall_val} mm)", f"Crop ({crop})"],
                    "synergy": "Severe Structural & Hydrological Vulnerability Loop",
                    "mechanism": (
                        "Sub-0.5% SOC destroys aggregate stability (slaking). Bare soil reaches >45°C "
                        "under solar irradiance, converting 50-65% of precipitation into evaporative runoff "
                        "and suffocating arbuscular mycorrhizal networks."
                    )
                })
                bottlenecks.append({
                    "factor": "Hydrological & Soil Carbon Colloidal Deficit",
                    "liebig_constraint": "Water holding capacity is paralyzed by absence of humic substances."
                })
                degradation_score += 40

        # Soil pH dysbiosis
        if ph is not None and (ph < 5.5 or ph > 8.2):
            nexus_insights.append({
                "variables": [f"Soil pH ({ph})", "Rhizosphere Microflora", "Nutrient Bioavailability"],
                "synergy": "Biogeochemical Ion Blockade",
                "mechanism": "Solubilizes toxic Al3+ (<5.5) or fixes orthophosphates (>8.2), arresting Rhizobium nodulation."
            })
            bottlenecks.append({
                "factor": "Ion Toxicity / Bioavailability Arrest",
                "liebig_constraint": "Root meristem elongation is chemically inhibited."
            })
            degradation_score += 25

        # Interventions synthesis
        interventions.append({
            "tier": "Tier 1: Immediate Rhizosphere & Moisture Foundation",
            "title": "Multi-Species Cover Crop Polyculture + AMF Inoculation",
            "protocol": "Vicia villosa (Hairy Vetch) + Secale cereale + Raphanus sativus (Daikon biological drilling) with Rhizophagus irregularis inoculant.",
            "mechanism": "Terminated with roller-crimper to create persistent mulch shield (-5°C topsoil cooling, +20-30% AWHC).",
            "citation": "FAO Soils Bulletin 44 (2021) & Lal (Science 2004)"
        })

        if soc is not None and soc < 1.0:
            interventions.append({
                "tier": "Tier 2: Permanent Carbon & Porous Sponge Matrix",
                "title": "Quenched Pyrolyzed Biochar Amendment (550°C)",
                "protocol": "8-10 t/ha biochar co-composted 1:4 with organic manure to saturate internal micro-porosity.",
                "mechanism": "Creates recalcitrant aromatic carbon sponge holding capillary water against vapor pressure deficit.",
                "citation": "Lehmann & Joseph (2015)"
            })

        interventions.append({
            "tier": "Tier 3: Landscape Heterogeneity & Pollinator Matrix",
            "title": "Stratified Alley Agroforestry & Floral Corridors",
            "protocol": "Reverse-phenology Faidherbia albida (10m x 10m) + native perennial floral buffer strips.",
            "mechanism": "Hydraulic lift provides subsoil water transfer; buffers ambient wind speed by 60-80%.",
            "citation": "IPCC AR6 WG2 Chapter 5 (2022) & IPBES (2019)"
        })

        # 5-Year Mathematical Pedotransfer Trajectory
        base_soc = soc if soc is not None else 0.5
        soc_trajectory = [round(base_soc + 0.3 * yr * (0.88 ** (yr * 0.2)), 2) for yr in [0, 1, 2, 3, 5]]

        # Hybrid RAG retrieval
        rag_query = f"{query} soil organic carbon rainfall {crop} biodiversity restoration"
        citations = self.rag.hybrid_search(rag_query, top_k=3)

        return {
            "executive_diagnosis": {
                "degradation_severity": f"{min(100, degradation_score)}%",
                "ecosystem_status": "Critically Compromised" if degradation_score > 60 else "Degraded Agro-Ecosystem",
                "summary": f"Multi-variable bottleneck diagnosed at the intersection of soil carbon ({base_soc}%), climate aridity ({rainfall_val}mm), and management disturbance."
            },
            "nexus_couplings": nexus_insights,
            "limiting_bottlenecks": bottlenecks,
            "recommended_interventions": interventions,
            "5_year_projections": {
                "soc_trajectory_pct": soc_trajectory,
                "projected_soc_gain": f"+{round(soc_trajectory[-1] - base_soc, 2)}%",
                "water_stored_liters_ha": round((soc_trajectory[-1] - base_soc) * 175000)
            },
            "literature_evidence": [
                {"title": c["title"], "authors": c["authors"], "year": c["year"], "doi": c["doi"], "rrf": c["provenance"]["rrf_score"]}
                for c in citations
            ]
        }

if __name__ == "__main__":
    reasoner = PythonNexusReasoner()
    test_state = {
        "soil_organic_carbon": 0.3,
        "soil_ph": 8.1,
        "annual_rainfall": 320,
        "crop": "monoculture wheat",
        "tillage": "conventional-inversion-tillage"
    }
    result = reasoner.evaluate(test_state, "monoculture wheat low rainfall")
    print("🌿 Python Multi-Metric Nexus Reasoning Result:")
    print(f"Severity: {result['executive_diagnosis']['degradation_severity']}")
    print(f"Projected 5-Yr SOC: {result['5_year_projections']['soc_trajectory_pct']} (Gain: {result['5_year_projections']['projected_soc_gain']})")
    print(f"Water Stored: +{result['5_year_projections']['water_stored_liters_ha']:,} L/ha")
    print("\nInterventions:")
    for i in result["recommended_interventions"]:
        print(f" - [{i['tier']}] {i['title']} (Ref: {i['citation']})")

File: server/engine/nexusReasoner.js
// Multi-Metric Causal Ecological Reasoning Engine
// Evaluates >= 3 environmental variables simultaneously to diagnose ecosystem collapse and synthesize non-obvious, evidence-grounded restoration protocols.

import { METRIC_DEFINITIONS } from './ecoMetrics.js';
import { causalGraphInstance } from './causalGraph.js';
import { INTERVENTIONS } from './interventionCatalog.js';
import { QuantitativeEcoModel } from './quantitativeModel.js';
import { ragEngineInstance } from '../knowledge/ragEngine.js';

export class MultiMetricNexusReasoner {
  /**
   * Performs deep multi-metric nexus reasoning on the given environmental state.
   * @param {Object} state - Normalized environmental state
   * @param {string} userQuery - Original user query or intent
   * @returns {Object} - Comprehensive scientific diagnostic dossier
   */
  static reason(state, userQuery = '') {
    // 1. Extract active environmental variables
    const activeMetrics = [];
    if (state.soil_organic_carbon !== undefined) activeMetrics.push('soil_organic_carbon');
    if (state.soil_ph !== undefined) activeMetrics.push('soil_ph');
    if (state.annual_rainfall !== undefined || state.rainfall !== undefined) activeMetrics.push('annual_rainfall');
    if (state.tillage_intensity !== undefined || state.tillage !== undefined) activeMetrics.push('tillage_intensity');
    if (state.crop_rotation_diversity !== undefined || state.crop !== undefined) activeMetrics.push('crop_rotation_diversity');
    if (state.chemical_input_intensity !== undefined) activeMetrics.push('chemical_input_intensity');
    if (state.habitat_fragmentation !== undefined) activeMetrics.push('habitat_fragmentation');
    if (state.shannon_diversity_index !== undefined || state.biodiversity !== undefined) activeMetrics.push('shannon_diversity_index');

    // 2. Multi-Variable Nexus Couplings Analysis (Evaluates >= 3 variables together)
    const nexusInsights = [];
    const limitingBottlenecks = [];
    let degradationScore = 0;

    // Numerical variables normalization
    const soc = typeof state.soil_organic_carbon === 'number' ? state.soil_organic_carbon : null;
    const ph = typeof state.soil_ph === 'number' ? state.soil_ph : null;
    const rainfallStr = (state.rainfall || '').toLowerCase();
    const isSemiArid = rainfallStr.includes('low') || rainfallStr.includes('semi-arid') || rainfallStr.includes('arid') || (typeof state.annual_rainfall === 'number' && state.annual_rainfall < 600);
    const cropStr = (state.crop || state.crop_rotation_diversity || '').toLowerCase();
    const isMonoculture = cropStr.includes('monoculture') || cropStr.includes('wheat') || cropStr.includes('single') || cropStr.includes('bare');
    const tillageStr = (state.tillage || state.tillage_intensity || '').toLowerCase();
    const isHighTillage = tillageStr.includes('conventional') || tillageStr.includes('inversion') || tillageStr.includes('high') || tillageStr.includes('plow');

    // Nexus Coupling 1: Soil Carbon (Low) x Semi-Arid Climate x Monoculture Crop (Triple Coupling)
    if ((soc !== null && soc < 1.0) || isSemiArid || isMonoculture) {
      if ((soc !== null && soc <= 0.5) && isSemiArid && isMonoculture) {
        nexusInsights.push({
          variables: ['Soil Organic Carbon (SOC 0.3%)', 'Semi-Arid Low Precipitation', 'Monoculture Wheat Cropping'],
          synergy_type: 'Severe Structural & Hydrological Vulnerability Loop',
          analysis: 'Low SOC (<0.5%) combined with monoculture wheat and low rainfall triggers a severe ecological feedback loop: absence of organic humus collodial matrix leads to topsoil aggregate breakdown (slaking). When sporadic convective rains occur, soil crusts rapidly, converting 50-65% of precipitation into evaporative loss and runoff instead of infiltration. Under high solar radiation, bare soil reaches temperatures >45°C, sterilizing surface mycorrhizal hyphae and starving pollinator/beneficial arthropod guilds.',
          severity: 'CRITICAL'
        });
        limitingBottlenecks.push({
          factor: 'Hydrological & Soil Carbon Collodial Deficit',
          liebig_constraint: 'Water holding capacity and aggregate stability are paralyzed by near-zero humic substances, rendering standard chemical inputs ineffective.'
        });
        degradationScore += 45;
      }
    }

    // Nexus Coupling 2: Soil pH Dysbiosis x Synthetic Input x Tillage
    if (ph !== null && (ph < 5.5 || ph > 8.2)) {
      nexusInsights.push({
        variables: ['Soil pH (' + ph + ')', 'Rhizosphere Microflora', 'Nutrient Bioavailability'],
        synergy_type: 'Biogeochemical Blockade',
        analysis: ph < 5.5
          ? `Acidic soil reaction (pH ${ph}) solubilizes toxic trivalent aluminum (Al³⁺), arresting root meristem elongation and completely inhibiting symbiotic Rhizobium nitrogenase enzyme activity. Consequently, natural biological N-fixation is arrested.`
          : `Alkaline/Sodic conditions (pH ${ph}) precipitate essential orthophosphates and micronutrients (Fe, Zn), while high exchangeable sodium disperses clay colloids, causing anoxic asphyxiation in earthworm and fungal communities.`,
        severity: 'HIGH'
      });
      limitingBottlenecks.push({
        factor: ph < 5.5 ? 'Aluminum Toxicity & Nitrogenase Arrest' : 'Sodic Dispersion & Micronutrient Fixation',
        liebig_constraint: 'Root elongation and microbial nodulation are chemically blocked.'
      });
      degradationScore += 25;
    }

    // Nexus Coupling 3: Tillage Disturbance x Fungal-Bacterial Ratio x Landscape Simplification
    if (isHighTillage || isMonoculture) {
      nexusInsights.push({
        variables: ['Tillage Regime', 'Fungal:Bacterial Ratio', 'Landscape Homogeneity'],
        synergy_type: 'Belowground Food Web & Aboveground Trophic Decoupling',
        analysis: 'Mechanical inversion tillage disrupts permanent earthworm vertical macropores and severs arbuscular mycorrhizal (AMF) networks. Concurrently, monoculture cropping eliminates continuous floral nectar sources, reducing wild bee (Apoidea) foraging density below the ecological threshold required for resilient cross-pollination.',
        severity: 'MODERATE_TO_HIGH'
      });
      degradationScore += 25;
    }

    // Default baseline degradation if parameters indicate stress
    if (degradationScore === 0) degradationScore = 30;

    // 3. Match Interventions from Scientific Catalog
    const matchedInterventions = [];
    
    // Check Multi-species cover crops
    if ((soc !== null && soc < 1.8) || isMonoculture || isHighTillage) {
      matchedInterventions.push(INTERVENTIONS.find(i => i.id === 'multi-species-cover-crop-mycorrhizal'));
    }

    // Check Biochar Pyrolysis
    if ((soc !== null && soc <= 1.0) || isSemiArid) {
      matchedInterventions.push(INTERVENTIONS.find(i => i.id === 'biochar-pyrolysis-co-composting'));
    }

    // Check Agroforestry Alleys
    if (isSemiArid || isMonoculture) {
      matchedInterventions.push(INTERVENTIONS.find(i => i.id === 'stratified-alley-agroforestry'));
    }

    // Check Floral Hedgerows
    matchedInterventions.push(INTERVENTIONS.find(i => i.id === 'perennial-floral-corridors-hedgerows'));

    // Check pH Rebalancing if pH is deviant
    if (ph !== null && (ph < 5.8 || ph > 7.8)) {
      matchedInterventions.push(INTERVENTIONS.find(i => i.id === 'ph-buffering-calcification-rebalancing'));
    }

    // 4. Retrieve Scientific Literature via Hybrid RAG
    const ragQuery = [
      userQuery,
      soc !== null ? `soil organic carbon ${soc}%` : '',
      isSemiArid ? 'semi-arid low rainfall drought resilience' : '',
      isMonoculture ? 'monoculture crop diversification agroforestry' : '',
      ph !== null ? `soil pH ${ph} buffering` : '',
      'biodiversity pollinator microbial biomass'
    ].filter(Boolean).join(' ');

    const retrievedEvidence = ragEngineInstance.search(ragQuery, { topK: 4 });

    // 5. Compute Quantitative Projections
    const quantitativeProjections = QuantitativeEcoModel.projectTrajectories(state, matchedInterventions);

    // 6. Trace Causal Downstream Feedbacks
    const causalTraces = causalGraphInstance.getDownstreamImpacts(
      activeMetrics.length > 0 ? activeMetrics : ['soil_organic_carbon', 'tillage_intensity']
    );

    return {
      executive_diagnosis: {
        degradation_severity_score: `${Math.min(100, degradationScore)}%`,
        ecosystem_state: degradationScore > 60 ? 'Critically Compromised Multifunctional State' : degradationScore > 35 ? 'Moderately Degraded Agro-Ecosystem' : 'Suboptimal Resilience',
        summary: `Ecosystem analysis reveals a multi-variable bottleneck driven by the intersection of ${activeMetrics.join(', ') || 'depleted soil carbon, climate stress, and agricultural simplification'}. Restoration requires coupled biogeochemical and spatial interventions rather than isolated single-variable amendments.`
      },
      nexus_multi_variable_couplings: nexusInsights,
      limiting_ecological_bottlenecks: limitingBottlenecks,
      recommended_interventions: matchedInterventions.filter(Boolean),
      quantitative_projections: quantitativeProjections,
      causal_feedback_traces: causalTraces.slice(0, 6),
      scientific_literature_evidence: retrievedEvidence,
      monitoring_bioindicators: [
        { indicator: 'Soil Microbial Respiration & Active Carbon (POXC)', frequency: 'Bi-annual', threshold: '>400 mg/kg target' },
        { indicator: 'Infiltration Rate (Single-Ring Infiltrometer)', frequency: 'Quarterly', threshold: '>50 mm/hour without ponding' },
        { indicator: 'Wild Pollinator Visitation Transect', frequency: 'Peak bloom', threshold: '>10 wild bees / 100 flowers / 10 min' },
        { indicator: 'Earthworm Burrow Abundance in Top 30cm', frequency: 'Spring/Autumn', threshold: '>120 individuals/m²' }
      ]
    };
  }
}
